bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -0,0 +1,216 @@
+    // TODO(ejona86): If the call fails for Auth reasons, this does not properly propagate info that
+    // would be in WWW-Authenticate, because it does not yet have access to the header.
+    return new CheckedForwardingCall<ReqT, RespT>(next.newCall(method)) {
+      @Override
+      protected void checkedStart(Listener<RespT> responseListener, Metadata.Headers headers)
+          throws Exception {
+        headers.put(AUTHORIZATION_HEADER_KEY, getHeader());
+        delegate().start(responseListener, headers);
+      }
+    };
+  }
+
+  public void asyncRefresh() {
+    if (!isRefreshing.get()) {
+      executor.execute(new Runnable() {
+        @Override
+        public void run() {
+          boolean requiresRefresh = false;
+          synchronized (isRefreshing) {
+            if (!isRefreshing.get()) {
+              isRefreshing.set(true);
+              requiresRefresh = true;
+            }
+          }
+          if (requiresRefresh) {
+            doRefresh();
+          }
+        }
+      });
+    }
+  }
+
+  /**
+   * Get the oauth2 AccessToken.
+   */
+  public String getHeader() throws IOException {
+    TokenCache tokenCache = getCachedHeader();
+    CacheState state = (tokenCache == null) ? CacheState.Expired : tokenCache.getCacheState();
+    if (state == CacheState.Expired) {
+      synchronized (isRefreshing) {
+        if (!isRefreshing.get()) {
+          isRefreshing.set(true);
+          doRefresh();
+        } else {
+          while (isRefreshing.get()) {
+            try {
+              isRefreshing.wait(250);
+            } catch (InterruptedException e) {
+              throw new IOException(e);
+            }
+          }
+        }
+      }
+      tokenCache = getCachedHeader();
+    } else if (state == CacheState.Stale) {
+      asyncRefresh();
+    } 
+    return tokenCache.header;
+  }
+  
+  private TokenCache getCachedHeader() throws IOException {
+    TokenCache token = tokenCache.get();
+    if (token != null && token.exception != null) {
+      throw token.exception;
+    }
+    return token;
+  }
+
+  /**
+   * Perform a credentials refresh.
+   */
+  private void doRefresh() {
+    try {
+      AccessToken newToken = credentials.refreshAccessToken();
+      if (newToken == null) {
+        // The current implementations of refreshAccessToken() throw an IOException or return
+        // a valid token. This handling is future proofing.
+        this.tokenCache.set(new TokenCache(new IOException("Could not load the token for credentials: " + credentials)));
+      } else {
+        long expiresTime = newToken.getExpirationTime().getTime();
+        String header = "Bearer " + newToken.getTokenValue();
+        long stateTime = expiresTime - TOKEN_STALENESS_MS;
+        // Block until refresh at this point.
+        long cacheExpiration = expiresTime - TOKEN_EXPIRES_MS;
+        this.tokenCache.set(new TokenCache(header, stateTime, cacheExpiration));
```

Comment:
stateTime -> staleTime

Also you could move the time logic inside the TokenCache constructor:
TokenCache(header, expiresTimeMs) {
  this.staleTime = expiresTimeMs - TOKEN_STALENESS_MS;
  this.expiresTime = expiresTimeMs - TOKEN_EXPIRES_MS